commcare_export/checkpoint.py

```diff
@@ -167,18 +222,84 @@ def list_checkpoints(self, limit=20):
         Don't filter by MD5 on purpose.
         """
         with session_scope(self.Session) as session:
-            query = session.query(Checkpoint)
+            query = self._filter_query(session.query(Checkpoint))
             if self.query:
                 query = query.filter(Checkpoint.query_file_name == self.query)
-            if self.project:
-                query = query.filter(Checkpoint.project == self.project)
-            if self.commcare:
-                query = query.filter(Checkpoint.commcare == self.commcare)
-            if self.key:
-                query = query.filter(Checkpoint.key == self.key)
-
             return query.order_by(Checkpoint.time_of_run.desc())[:limit]
 
+    def _filter_query(self, query):
+        if self.project:
+            query = query.filter(Checkpoint.project == self.project)
+        if self.commcare:
+            query = query.filter(Checkpoint.commcare == self.commcare)
+        if self.key:
+            query = query.filter(Checkpoint.key == self.key)
+        return query
+
+    def get_latest_checkpoints(self):
+        """Returns the latest checkpoint for each table filtered by the fields set in the manager:
+        * query_md5
+        * project
+        * commcare
+        * key
+        """
+        with session_scope(self.Session) as session:
+            window_func = func.row_number().over(
+                partition_by=Checkpoint.table_name, order_by=Checkpoint.time_of_run.desc()
+            ).label("row_number")
+            inner_query = self._filter_query(session.query(Checkpoint, window_func))
+            inner_query = inner_query.filter(Checkpoint.query_file_md5 == self.query_md5)
+            inner_query = inner_query.filter(Checkpoint.table_name.isnot(None)).subquery()
+
+            query = session.query(Checkpoint).select_entity_from(inner_query)\
+                .filter(inner_query.c.row_number == 1)\
+                .order_by(Checkpoint.table_name.asc())
+            return list(query)
+
     def update_checkpoint(self, run):
         with session_scope(self.Session) as session:
             session.merge(run)
+
+    def _validate_tables(self):
+        if not self.table_names:
+            raise Exception("Not tables set in checkpoint manager")
+
+
+class CheckpointManagerWithSince(object):
+    def __init__(self, manager, since):
+        self.manager = manager
+        self.since_param = since
+
+    def set_checkpoint(self, checkpoint_time, is_final=False):
+        if self.manager:
```

Comment:
Are there known scenarios where `set_checkpoint()` would be called when `self.manager is None`? If not, what do you think of raising an exception here rather than returning silently?

Prior to this PR, it looks like a call to `set_checkpoint()` always set a checkpoint, but this makes it possible that it would be a no-op. That seems like it could be bad (if there was a bug that caused `self.manager` to be false-y).

Reply:
I'm not blocking on this, but it sounds like the checkpoint manager is maybe doing more than one thing[<sup>1</sup>](#do-one-thing) if it sometimes is used to set checkpoints and other times is only the source of the `since` value.


<sup><a name="do-one-thing">1</a></sup> ref. "Do One Thing" concept from [Clean Code](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882), which we're reading in book club right now.

Reply:
I have already applied this concept:
* `CheckpointManager` is solely responsible for managing checkpoints.
* `CheckpointManagerWithSince` is a container for the `since` value and checkpoint manager specific to the current table export.
* `CheckpointProvider` is responsible for creating new the creating a new `CheckpointManagerWithSince` when requested with the `since` value and the checkpoint manager.

The `since` value and the checkpoints are inextricably tied together because unless specified at the command line the `since` value comes from the checkpoints which is why it makes sense to be part of the checkpoint manager: https://github.com/dimagi/commcare-export/pull/121/files#diff-a29d76e0f5317a0f68900178498d8141R314

Having said all that I'm open to suggestions for improvement. I did got through several iterations with this and originally didn't have the `CheckpointManagerWithSince`. I introduced that class so that the checkpoint manager with tables set would only live inside that class which in turn would only live inside the context of the of `Bind`.

Reply:
You highlighted what I'm getting at by using the word "and" in the description of `CheckpointManagerWithSince`:

> a container for the `since` value _and_ checkpoint manager specific to the current table export.

(emphasis added)

Note: previously I had referenced "Do One Thing," a concept Uncle Bob uses when talking about functions, when I actually meant to reference "Single Responsibility Principle," which he uses when talking about classes. Here's a relevant quote from the chapter on Classes near where he's talking about the Single Responsibility Principle:

> The name of a class should describe what responsibilities it fulfills. In fact, naming is probably the first way of helping determine class size. If we cannot derive a concise name for a class, then it’s likely too large. The more ambiguous the class name, the more likely it has too many responsibilities. For example, class names including weasel words like `Processor` or `Manager` or `Super` often hint at unfortunate aggregation of responsibilities. We should also be able to write a brief description of the class in about 25 words, without using the words “if,” “and,” “or,” or “but.”

I'm going to let this one go. The PR fixes the issue that prompted the change, and I don't have enough context here to easily suggest how to improve it.

Reply:
I'm aware of the principle but I disagree that it applies here. It would be like applying it to a DB access class that reads "and" writes.

As I already said, the `since` value is inextricably tied to the checkpoints. If I rename it like this does it make it clearer for you:

```python
class CheckpointInterface(object):
    def get_initial_checkpoint_value(self)
        pass

    def set_checkpoint(self, value)
        pass
```

Reply:
Does `get_initial_checkpoint_value()` return what used to be `since_param`? Will that interface be implemented by both `CheckpointManagerWithSince` and `CheckpointManager`? I guess it would be helpful to see the implementation(s) of that interface.

Reply:
Yes, it would return the same value. I'll try and find some time this week to refactor these classes.